### PR TITLE
feat: streaming debug logfile

### PR DESCRIPTION
**Pull request by lukekarrys**
This decouples the log file writing from the terminal logging. We now
open an append only file at the start of the process and stream logs to
it. We still only display the log file message in timing mode or if
there is an error, but the file is still written regardless.

All logging now goes through `proc-log` and this is the first step to
removing `npmlog`. For now `npmlog` is still used for the terminal
logging but with a shim in front of it to make it easier to test and
use in conjunction with `proc-log`. Ref: npm/statusboard#366

This also refactors many of the tests to always use an explicit
`t.testdir` for their cache since the file is opened on each `new Npm()`.
Tests are also refactored to use more of `MockNpm` with behavior to
add config items and load `npm` if necessary. A new fixture `mockGlobals`
was also added to make much of this more ergonomic. Ref: npm/statusboard#410

Closes npm/statusboard#411
Closes npm/statusboard#367

### CODE REVIEW

File: lib/commands/exec.js [hunk 1]
```diff
@@ -77,14 +77,14 @@ class Exec extends BaseCommand {
...
       locationMsg,
-      log,
       globalBin,
       output,
       packages,
       path,
       runPath,
       scriptShell,
       yes,
+      log,
```
Comment (wraithgar) [inline lib/commands/exec.js]: why was this moved out of alphabetical order?

File: lib/commands/init.js [hunk 2]
```diff
@@ -114,7 +114,6 @@ class Init extends BaseCommand {
       color,
       localBin,
       locationMsg,
-      log,
```
Comment (wraithgar) [inline lib/commands/init.js]: libnpmexec still needs this, no?

File: lib/commands/exec.js (hunk 1 above)
Comment (Branscott1780) [reply lib/commands/exec.js]: Because the global name must be placed first .in order to get function of the contributor.

Comment (Branscott1780) [reply lib/commands/exec.js]: So the path could be a constant for the shell

File: lib/npm.js [hunk 3]
```diff
@@ -258,18 +289,35 @@ class Npm extends EventEmitter {
...
-    return log
+  get unfinishedTimers () {
+    return this.#timers.unfinished
+  }
+
+  get finishedTimers () {
+    return this.#timers.finished
+  }
+
+  get started () {
```
Comment (wraithgar) [inline lib/npm.js:315]: is this too ambiguous?

File: lib/utils/display.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    if (unicode) {
+      log.enableUnicode()
+    } else {
+      log.disableUnicode()
+    }
+
+    // Progress is on stderr so see if term supports that
+    const stderrTTY = process.stderr.isTTY
```
Comment (wraithgar) [inline lib/utils/display.js]: future concern: these are prime candidates for derived config values

File: lib/utils/display.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    this[_logHandler](...args)
+  }
+
+  [_logHandler] = (level, ...args) => {
+    try {
+      this[_log](level, ...args)
+    } catch (ex) {
+      try {
+        // if it crashed once, it might again!
+        this[_npmlog]('verbose', `attempt to log ${inspect(args)} crashed`, ex)
```
Comment (wraithgar) [inline lib/utils/display.js:84]: This feels like extremely legacy overly-optimized behavior and I wonder if we need this still

File: tap-snapshots/test/lib/utils/error-message.js.test.cjs [hunk 6]
```diff
@@ -255,7 +255,7 @@ Object {
   "summary": Array [
     Array [
       "notsup",
-      "Unsupported platform for lodash@1.0.0: wanted {\\"os\\":\\"!yours,mine\\",\\"arch\\":\\"x867,x5309\\"} (current: {\\"os\\":\\"posix\\",\\"arch\\":\\"x64\\"})",
+      "Unsupported platform for lodash@1.0.0: wanted {/"os/":/"!yours,mine/",/"arch/":/"x867,x5309/"} (current: {/"os/":/"posix/",/"arch/":/"x64/"})",
```
Comment (wraithgar) [inline tap-snapshots/test/lib/utils/error-message.js.test.cjs:258]: this is odd, right?

**Review by wraithgar — APPROVED** [review]

**Comment by Branscott1780** [comment]
good luck on the new iPhone
